src/main/java/at/hannibal2/skyhanni/features/dungeon/CroesusChestTracker.kt

```diff
@@ -0,0 +1,250 @@
+    private val croesusEmptyPattern by repoGroup.pattern("empty", "§cNo treasures!")
+    private val kismetPattern by repoGroup.pattern("kismet.reroll", "§aReroll Chest")
+    private val kismetUsedPattern by repoGroup.pattern("kismet.used", "§aYou already rerolled a chest!")
+
+    private val floorPattern by repoGroup.pattern("chest.floor", "§7Tier: §eFloor (?<floor>[IV]+)")
+    private val masterPattern by repoGroup.pattern("chest.master", ".*Master.*")
+
+    private val keyUsedPattern by repoGroup.pattern("chest.state.keyused", "§aNo more Chests to open!")
+    private val openedPattern by repoGroup.pattern("chest.state.opened", "§8Opened Chest:.*")
+    private val unopenedPattern by repoGroup.pattern("chest.state.unopened", "§8No Chests Opened!")
+
+    private val kismetSlotId = 50
+    private val emptySlotId = 22
+    private val frontArrowSlotId = 53
+    private val backArrowSlotId = 45
+
+    private val kismetInternalName = "KISMET_FEATHER".asInternalName()
+
+    private var inCroesusInventory = false
+    private var croesusEmpty = false
+    private var croesusPageNumber = 0
+    private var pageSwitchable = false
+
+    private var chestInventory: DungeonChest? = null
+
+    private var currentRunIndex = 0
+
+    private var kismetAmountCache = 0
+
+    @SubscribeEvent(priority = EventPriority.LOW)
+    fun onBackgroundDrawn(event: GuiContainerEvent.BackgroundDrawnEvent) {
+        if (!LorenzUtils.inSkyBlock) return
+        if (!SkyHanniMod.feature.dungeon.croesusUnopenedChestTracker) return
+
+        if (inCroesusInventory && !croesusEmpty) {
+            for ((run, slot) in InventoryUtils.getItemsInOpenChest()
+                .mapNotNull { slot -> runSlots(slot.slotIndex, slot) }) {
+
+                if (run.floor == null) return
+
+                val state = run.openState ?: OpenedState.UNOPENED
+
+                if (state != OpenedState.KEY_USED) {
+                    slot highlight if (state == OpenedState.OPENED) LorenzColor.DARK_AQUA else LorenzColor.DARK_PURPLE
+                }
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onInventoryOpen(event: InventoryFullyOpenedEvent) {
+        if (!LorenzUtils.inSkyBlock) return
+        if (SkyHanniMod.feature.dungeon.croesusUnopenedChestTracker || config.kismet) croesusPattern.matchMatcher(event.inventoryName) {
+            inCroesusInventory = true
+            pageSwitchable = true
+            croesusEmpty = croesusEmptyPattern.matches(event.inventoryItems[emptySlotId]?.name)
+            if (event.inventoryItems[backArrowSlotId]?.item != Items.arrow) {
+                croesusPageNumber = 0
+            }
+
+            if (croesusEmpty) {
+                croesusChests?.forEach {
+                    it.setValuesNull()
+                }
+                return
+            }
+
+            for ((run, item) in event.inventoryItemsWithNull.mapNotNull { (key, value) -> runSlots(key, value) }) {
```

Comment:
why do we need this "with null" logic? cant we just assume every slot index not in the map is null?

Reply:
Since it also needs to check if an item is not present ( null ). So it can clear the chest in the storage.

Added a comment to explain this.